Add support for cookies with "SameSite" option.

ADmad opened the pull request:
https://www.owasp.org/index.php/SameSite

I can add / update tests if the changes seem reasonable.

Todo:
- [x]  Add / updates tests
- [x] Update `Http\Client\Response`.

ADmad committed:
Add support for cookies with "SameSite" option.

markstory commented on src/Http/Cookie/Cookie.php:
```diff
@@ -138,13 +148,50 @@ public function __construct(
+     *  (Keys must be lowercase).
+     * @return static
+     */
+    public static function create(string $name, $value, array $options = [])
```
:+1: This is a bit simpler than the constructor for basic usage.

markstory commented on src/Http/Cookie/Cookie.php:
```diff
@@ -535,6 +624,26 @@ public function isExpanded(): bool
+    /**
+     * @inheritDoc
+     */
+    public function getOptions(): array
```
How do you see this being used? Is it for tests, or as part of an application?

ADmad replied:
It is used to gets args for new signature of setcookie() call in ResponseEmitter.

markstory commented on src/Http/ResponseEmitter.php:
```diff
@@ -264,16 +269,28 @@ protected function emitCookies(array $cookies): void
+            }
+
+            if (PHP_VERSION_ID >= 70300) {
+                setcookie($name, $value, $cookie['options']);
```
Should we raise an exception if people try and use samesite cookies when they are on PHP<7.3?

ADmad replied:
I'll add an exception.

garas replied:
Could it generate header with `Cookie::toHeaderValue()` and send it using `header()` instead?

Exceptions might be missed if development and production uses different PHP version.

ADmad replied:
I guess that's an option too.

ADmad replied:
Directly switching to `header()` for PHP < 7.3 would making testing a pain and we would need different set of tests for different PHP versions.

Instead we can use the namespaced `Cake\Http\setcookie()` currently only used in tests in `ResponseEmitter` too which can internally either call PHP's `setcookie()` or `header()`.

garas replied:
I would use same `header()` always independently of PHP.

ADmad committed:
Add CookieInterface::toArray() method.

ADmad committed:
Fix array keys

ADmad committed:
Update Reponse tests related to cookies.

ADmad committed:
Update CsrfProtectionMiddleware tests related to cookies.

ADmad commented:
There are few other issues which I want to address:

1. One would expect `Cake\Http\Response::getCookie()` and `Cake\Http\Client\Response::getCookie()` to return similar data but they don't. The former returns an array representation of cookie with keys `name`, `value`, `path` etc. while the latter just return the cookie value.
2. `Cake\Http\Client\Response::getCookieData()` returns cookie data array similar `Cake\Http\Response::getCookie()` but there is discrepancy in key names: `expires` vs `expire`, `httponly` vs `httpOnly`.  
Also for `Http\Response` the `expire` key is timestamp while for `Http\Client\Response` it's a formatted string. 
Is this comment `This method is compatible with older client code that expects date strings instead of timestamps.` for `Http\Cllient\Reponse::convertCookieToArray()`  still relevant?

So my proposal is:

1. Make `getCookie()` method of both response classes return `Cookie` instance instead of array. Using the object methods is lot more convenient dealing using array keys which can easily have typos (especially considering the discrepancy in key name).
2. Add a `CookieInterface::toArray()` method which returns array of form
`['name' => '', 'value' => '', 'options' => ['path' => '', ...]]`. The `convertCookieToArray()` methods of both response classes could then be dropped.

ADmad committed:
Add Cookie::createFromHeaderString() method.

ADmad committed:
Avoid null expires.

setcookie() expects expires to be int.

ADmad committed:
Fix error reported by psalm.

ADmad commented:
Either I have made a silly mistake which I cannot spot or PHP is trolling me. Using the alternate `setcookie()` signature with 3rd argument as array for PHP 7.3+ throws error:

`ArgumentCountError: Too few arguments to function Cake\Http\setcookie(), 3 passed in /home/runner/work/cakephp/cakephp/src/Http/ResponseEmitter.php on line 228 and at least 5 expected`

garas commented:
Troll lives in https://github.com/cakephp/cakephp/blob/4.x/tests/TestCase/Http/server_mocks.php

ADmad commented:
Thanks, I got the hint when i saw `Cake\Http\setcookie()` as function names instead of `setcookie()` but it's too late in the day to go troll hunting, so I'll take care of it tomorrow :stuck_out_tongue:.

ADmad committed:
Fix mocked setcookie()'s signature.

Also update option key name.

ADmad committed:
Add tests.

ADmad committed:
Remove code duplication

ADmad committed:
More code cleanup

ADmad commented:
I added a new static method `Cookie::createFromHeaderString()` to avoid the duplicate string parsing code in `ResponseEmitter` and `Client/Response`.

Only dealing with the `SameSite` for PHP < 7.3 and changing the return type of `getCookie()` methods of the two response classes remains.

ADmad commented:
I am a bit sure whether `none` should also be allowed as a valid value for `SameSite`. [OWASP](https://www.owasp.org/index.php/SameSite) only mentions `lax` and `strict` as valid values but I have seen various articles which seems to indicate that chrome can have `none` too.

ADmad committed:
Ignore invalid samesite value when parsing header.

https://tools.ietf.org/html/draft-west-first-party-cookies-07#section-4.1

ADmad committed:
Allow using "SameSite" attribute for PHP < 7.3.

markstory commented on src/Http/ResponseEmitter.php:
```diff
@@ -213,68 +214,47 @@ protected function emitHeaders(ResponseInterface $response): void
-                    $key = $part;
-                    $value = true;
-                }
+        header('Set-Cookie: ' . $cookie);
```
Have you checked that this will play nicely with sessions and setcookie()?

How will this work in integration tests?

garas replied:
You need to pass in `false` as the second argument to allow multiple `Set-Cookie` headers.

ADmad replied:
> Have you checked that this will play nicely with sessions and setcookie()?

IDK, I just followed @garas's suggestion. If it doesn't work then users of PHP 7.3 just won't be able to use `SameSite`.

> How will this work in integration tests?

There are no test failures :) 

> You need to pass in false as the second argument to allow multiple Set-Cookie headers.

Fixed

ADmad replied:
Seems the `SameSite` restriction for session cookie can be handled this way https://github.com/delight-im/PHP-Cookie/blob/master/src/Session.php#L146.

markstory replied:
I'm not concerned about SameSite session cookies, more that using the `set-cookie` header can cause sessions to not work. This commit 7ebc2614fd9e32beec80a2c5dee9cf73f288bb1e has a removal in the `ResponseTransformer` that was a workaround for session cookies not being emitted.

ADmad replied:
Okay, I really don't want to go down that rabbit hole for now. I'll revert the use of `header()` and simply throw exception is samesite is set for php 7.2.

We can look into supporting samesite attribute for PHP 7.2 at a later date.

markstory commented on tests/TestCase/Http/Client/ResponseTest.php:
```diff
@@ -354,13 +354,13 @@ public function testGetCookies()
-        $this->assertTrue($result['httponly']);
-        $this->assertTrue($result['secure']);
+        $this->assertTrue($result['options']['httponly']);
+        $this->assertTrue($result['options']['secure']);
```
This options key and renaming expire to expires is going to break applications. Could we keep this interface the same and add in `samesite`?

ADmad committed:
Make CookieInterface::toArray() return a flat array.

ADmad committed:
Fix header() call to avoid overwriting previous cookie

ADmad commented on src/Http/Response.php:
```diff
@@ -1295,34 +1293,12 @@ public function getCookies(): array
         $cookies = $this->_cookies;
         foreach ($cookies as $cookie) {
-            $out[$cookie->getName()] = $this->convertCookieToArray($cookie);
+            $out[$cookie->getName()] = $cookie->toArray();
```
I have changed back to flat array without the nesting for `options` key.

ADmad commented on src/Http/Cookie/Cookie.php:
```diff
@@ -535,6 +701,38 @@ public function isExpanded(): bool
+        return [
+            'name' => $this->name,
+            'value' => $this->getScalarValue(),
+            'options' => $this->getOptions(),
```
I have changed back to flat array without the nesting for `options` key. But I really want to keep the `expire` to `expires` rename for consistency.

markstory replied:
Would outputting both be an option?

ADmad replied:
If we output old keys too then even those staring fresh with 4.0 could inadvertently use the old keys and we wouldn't be able to do this clean up even in 5.0.

ADmad committed:
Don't use header() to set cookies.

Doing so has caused problems with setting session cookie in the past.
Instead use a hack to set "SameSite" attribute for PHP 7.2.

markstory commented on src/Http/ResponseEmitter.php:
```diff
@@ -227,34 +227,36 @@ protected function emitCookies(array $cookies): void
+        if ($sameSite !== null) {
+            // Temporary hack for PHP 7.2 to set "SameSite" attribute
+            // https://stackoverflow.com/questions/39750906/php-setcookie-samesite-strict
+            $path .= '; samesite=' . $sameSite;
```
Neat solution.

ADmad committed:
Avoid unnecessary iterator to array conversion.

ADmad commented:
This is ready for final review.

@markstory The only non BC change is change of key names in array returned by `Http\Response::getCookie()`. But I think the use cases for reading back cookies meant to be sent with response would be pretty rare. Usually those cookies are just set and emitted by server.

ADmad commented:
The `Static Analysis` job failure is unrelated and fixed on `4.x`.

markstory commented on src/Http/Cookie/Cookie.php:
```diff
@@ -535,6 +701,37 @@ public function isExpanded(): bool
+        return [
+            'name' => $this->name,
+            'value' => $this->getScalarValue(),
+        ] + $this->getOptions();
```
:clap:

markstory commented on tests/TestCase/Http/Middleware/CsrfProtectionMiddlewareTest.php:
```diff
@@ -278,10 +278,10 @@ public function testConfigurationCookieCreate()
         $this->assertSame('/dir/', $cookie['path'], 'session path.');
         $this->assertTrue($cookie['secure'], 'cookie security flag missing');
-        $this->assertTrue($cookie['httpOnly'], 'cookie httpOnly flag missing');
+        $this->assertTrue($cookie['httponly'], 'cookie httpOnly flag missing');
```
These key renames should be added to the migration guide. I don't think rector is going to help us here as we can't easily discern valid replacements.

ADmad replied:
I'll make a PR for the migration guide.
